litellm/integrations/weave/weave_otel.py

```diff
@@ -0,0 +1,338 @@
+WEAVE_CLOUD_ENDPOINT = "https://trace.wandb.ai/otel/v1/traces"
+
+
+class WeaveOtelLogger(OpenTelemetry):
+    """
+    Weave (W&B) OpenTelemetry Logger for LiteLLM.
+
+    Sends LLM traces to Weave via the OpenTelemetry Protocol (OTLP).
+
+    Environment Variables:
+        WANDB_API_KEY: Required. Weights & Biases API key for authentication.
+        WEAVE_PROJECT_ID: Required. Project ID in format <entity>/<project_name>.
+        WEAVE_HOST: Optional. Custom Weave host URL. Defaults to cloud endpoint.
+
+    Usage:
+        litellm.callbacks = ["weave_otel"]
+
+    Reference:
+        https://docs.wandb.ai/weave/guides/tracking/otel
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _maybe_log_raw_request(
+        self, kwargs, response_obj, start_time, end_time, parent_span
+    ):
+        """
+        Override to skip creating the raw_gen_ai_request child span.
+
+        For Weave, we only want a single span per LLM call. The parent span
+        already contains all the necessary attributes, so the child span
+        is redundant.
+        """
+        pass
+
+    @staticmethod
+    def set_weave_otel_attributes(span: Span, kwargs, response_obj):
+        """
+        Sets OpenTelemetry span attributes for Weave observability.
+        Uses the same attribute setting logic as other OTEL integrations for consistency.
+        """
+        _utils.set_attributes(span, kwargs, response_obj, WeaveLLMObsOTELAttributes)
+
+        # Set Weave-specific attributes
+        WeaveOtelLogger._set_weave_specific_attributes(
+            span=span, kwargs=kwargs, response_obj=response_obj
+        )
+
+    @staticmethod
+    def _extract_weave_metadata(kwargs: dict) -> dict:
+        """
+        Extracts Weave metadata from the standard LiteLLM kwargs structure.
+
+        Reads kwargs["litellm_params"]["metadata"] if present and is a dict.
+        """
+        litellm_params = kwargs.get("litellm_params", {}) or {}
+        metadata = litellm_params.get("metadata")
+        if not isinstance(metadata, dict):
+            metadata = {}
+        return metadata
+
+    @staticmethod
+    def _set_metadata_attributes(span: Span, metadata: dict):
+        """Helper to set metadata attributes from mapping."""
+        from litellm.integrations.arize._utils import safe_set_attribute
+        from litellm.litellm_core_utils.safe_json_dumps import safe_dumps
+
+        # Weave-specific attributes
+        weave_mapping = {
+            "thread_id": WeaveSpanAttributes.THREAD_ID,
+            "is_turn": WeaveSpanAttributes.IS_TURN,
+            "display_name": WeaveSpanAttributes.DISPLAY_NAME,
+        }
+
+        # Standard attributes recognized by Weave
+        standard_mapping = {
+            "trace_user_id": WeaveSpanAttributes.TRACE_USER_ID,
+            "user_id": WeaveSpanAttributes.TRACE_USER_ID,
+            "session_id": WeaveSpanAttributes.SESSION_ID,
+        }
```

Comment:
What do these correspond to? We don't reserve these fields.

Reply:
Related to comment below